keys.go

```diff
@@ -0,0 +1,174 @@
+package matrixfederation
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/sha256"
+	"crypto/tls"
+	"encoding/json"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strings"
+)
+
+// ServerKeys are the ed25519 signing keys published by a matrix server.
+// Contains SHA256 fingerprints of the TLS X509 certificates used by the server.
+type ServerKeys struct {
+	Raw             []byte     `json:-`             // Copy of the raw JSON for signature checking.
+	ServerName      string     `json:"server_name"` // The name of the server.
+	TLSFingerprints []struct { // List of SHA256 fingerprints of X509 certificates.
+		SHA256 Base64String `json:"sha256"`
+	} `json:"tls_fingerprints"`
+	VerifyKeys map[string]struct { // The current signing keys in use on this server.
+		Key Base64String `json:"key"` // The public key.
+	} `json:"verify_keys"`
+	ValidUntilTS  uint64              `json:"valid_until_ts"` // When this result is valid until in milliseconds.
+	OldVerifyKeys map[string]struct { // Old keys that are now only valid for checking historic events.
+		Key       Base64String `json:"key"`        // The public key.
+		ExpiredTS uint64       `json:"expired_ts"` // When this key stopped being valid for event signing.
+	} `json:"old_verify_keys"`
+}
+
+// FetchKeysDirect fetches the matrix keys directly from the given address.
+// Optionally sets a SNI header if ``sni`` is not empty.
+// Returns the server keys and the state of the TLS connection used to retrive them.
+func FetchKeysDirect(serverName, addr, sni string) (*ServerKeys, *tls.ConnectionState, error) {
+	// Create a TLS connection.
+	tcpconn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer tcpconn.Close()
+	tlsconn := tls.Client(tcpconn, &tls.Config{ServerName: sni})
+	err = tlsconn.Handshake()
+	if err != nil {
+		return nil, nil, err
+	}
+	connectionState := tlsconn.ConnectionState()
+
+	// Write a GET /_matrix/key/v2/server down the connection.
+	requestURL := "matrix://" + serverName + "/_matrix/key/v2/server"
+	request, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	request.Header.Set("Connection", "close")
+	if err = request.Write(tlsconn); err != nil {
+		return nil, nil, err
+	}
+
+	// Read the 200 OK from the server.
+	response, err := http.ReadResponse(bufio.NewReader(tlsconn), request)
+	if err != nil {
```

Comment:
Missing `defer response.Close()`. If this is intentional (because of interesting side effects on `connectionState`?) then please comment why you're not closing this response.